### PR TITLE
Make stacks composable

### PR DESCRIPTION
**Pull request by akshayjshah**
This PR sketches out one approach to making stacks composable. It introduces a `Group` type (bikeshed away), that let modules transparently provide multiple types or offer an all-in-one starter.

### CODE REVIEW

File: stack.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		} else {
+			ret = append(ret, f)
+		}
+	}
+	return ret
+}
+
+// Constructors composes multiple constructors into a single Provide-able
+// object.
+type Constructors []interface{}
```
Comment (breerly) [inline stack.go]: Instead of calling this `Constructors`, I'd rather have an explicit name like `ConstructorGroup`.

File: stack.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+func (c Constructors) stack() []interface{} {
+	return []interface{}(c)
+}
+
+// Starters composes multiple functions into a single Start-able object.
+type Starters []interface{}
```
Comment (breerly) [inline stack.go]: Hmmm... this name is rough.... maybe... `InvokeGroup`?

Comment (akshayjshah) [inline stack.go]: Sure, sounds good.

File: stack.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		} else {
+			ret = append(ret, f)
+		}
+	}
+	return ret
+}
+
+// A Group composes multiple constructors or invokable functions into a single
+// object.
+type Group []interface{}
```
Comment (alsamylkin) [inline stack.go]: Out of curiosity, why do we need this type?

Comment (abhinav) [inline stack.go]: I think the idea is to make it opt-in rather than treating all slices as lists of constructors.

Comment (akshayjshah) [inline stack.go]: Yep, that's it. We can go a little further and do 

```
type Group interface {
   stack() []interface{}
}

func NewGroup(...interface{}) Group
```

which gives us a little wiggle room. We could, for example, add `NewNamedGroup(string, ...interface{})` later on.

**Comment by glibsm** [comment]
Shall we hold off on this until after beta4? We're pretty close to cutting it (even though we know there are some of these things we want to get done for future-proofing).

That would also remove some of the pressure to merge merge merge, as the train is about to leave.

Thoughts?

**Comment by breerly** [comment]
That makes sense to me G

**Comment by akshayjshah** [comment]
Sure, works for me.

File: app_test.go [hunk 4]
```diff
@@ -79,6 +79,29 @@ func TestApp(t *testing.T) {
...
+		}
+		biz := func(s2 *type2, s3 *type3) error {
+			count++
+			return nil
+		}
+		s := New()
+		s.Provide(Group{new1, new2, new3})
+		require.NoError(t, s.Start(context.Background(), biz))
+		assert.Equal(t, 3, count)
+	})
```
Comment (abhinav) [inline app_test.go]: Should we test for nested stacks?

Comment (akshayjshah) [inline app_test.go]: Sure, will add.

**Comment by glibsm** [comment]
Let me make sure I understand this correctly. 

Lets say I have `m1.Module` and `m2.Module` in packages `m1`, and `m2` respectively and I want to add a convenience thing that will combine both of them into a stack together.

Can I wrote an `m3.Modules` helper that will look like this?

```
package m3
func Module() fx.Group {
    return fx.Group{m1.Module, m2.Module}
}
```

**Comment by akshayjshah** [comment]
```
package m3

var Module = fx.Group{
  m1.Module,
  m2.Module,
}
```

**Comment by glibsm** [comment]
:mind_blown:

**Comment by breerly** [comment]
The more I look at it, the more I like `fx.Group`.

This API is growing on me quickly.

**Comment by breerly** [comment]
Although I wonder if this should go directly in dig, along with `dig.In` and `dig.Out`...

**Comment by abhinav** [comment]
I think we can get away with starting with it in fx. If/when we port the feature to dig natively, fx can simply delegate to it.

File: app_test.go (hunk 4 above)
Comment (akshayjshah) [inline app_test.go]: done.

File: stack.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+type stack interface {
+	stack() []interface{}
+}
+
+func unstack(funcs []interface{}) []interface{} {
```
Comment (glibsm) [inline stack.go]: nit: ungroup? To keep the terminology consistent

Comment (akshayjshah) [inline stack.go]: sure, can do.

**Comment by breerly** [comment]
Instead of `fx.Group`, how about `fx.Stack` - I want to keep the stack terminology around, and it feels smoother that fxstack *is* an fx.Stack.

Also, I think we've already started to ingrain "stack" in our terminology - I worry introducing "group" will actually move us a step away from clear and concise understandability.

**Comment by alsamylkin** [comment]
Stack has a very different meaning when we are talking about collections :)

**Comment by glibsm** [comment]
We're mixing and matching concepts here.

Stack is something that composes multiple modules together, and gets a change to affect startup, etc.

Group is something that ties multiple modules together *to be provided to the stack* and make the "break the glass" scenarios a lot easier, as well as tying together multiple constructors logically.

**Comment by breerly** [comment]
that feels wrong - we should shoot for a single concept

**Comment by breerly** [comment]
Let's discuss these concepts on a ca

**Comment by breerly** [comment]
Call*

**Comment by akshayjshah** [comment]
Sure, sounds good.

**Pull request closed without merging**